sdk/core/Azure.Core/src/ResourceIdentifier.cs

```diff
@@ -13,168 +13,208 @@ namespace Azure.Core
-            : this(parent, new ResourceType(providerNamespace, resourceTypeName), resourceName, true)
-        {
-        }
-
-        private void Init(ResourceIdentifier? parent, ResourceType resourceType, string resourceName)
-        {
-            if (parent is not null)
-            {
-                Provider = parent.Provider;
-                SubscriptionId = parent.SubscriptionId;
-                Location = parent.Location;
-                ResourceGroupName = parent.ResourceGroupName;
-            }
-
-            if (resourceType == SubscriptionResourceType)
+            if (resourceId == RootStringValue)
             {
-                if (!Guid.TryParse(resourceName, out _))
-                    throw new ArgumentOutOfRangeException(nameof(resourceName), $"The GUID for subscription is invalid {resourceName}.");
-                SubscriptionId = resourceName;
+                Name = String.Empty;
+                ResourceType = new ResourceType(ResourceType.Tenant);
+                return;
             }
 
-            if (resourceType.GetLastType() == LocationsKey)
-                Location = resourceName;
+            if (!remaining.StartsWith(SubscriptionStart.AsSpan()) && !remaining.StartsWith(ProviderStart.AsSpan()))
+                throw new ArgumentOutOfRangeException(nameof(resourceId), $"The ResourceIdentifier must start with {SubscriptionStart} or {ProviderStart}.");
 
-            if (resourceType == ResourceGroupResourceType)
-                ResourceGroupName = resourceName;
+            remaining = remaining.Slice(1);
 
-            if (resourceType == ProviderResourceType)
-                Provider = resourceName;
+            //we know we at least have 1 ResourceIdentifier in the tree here
+            ResourceIdentifierParts nextParts;
+            ReadOnlySpan<char> parentResourceType = ResourceType.Tenant.AsSpan();
+            do
+            {
+                _parentIndex = resourceId.Length - remaining.Length - 1;
 
-            if (parent is null)
-                _stringValue = RootStringValue;
-            Parent = parent;
-        }
+                //continue to get the next ResourceIdentifier in the tree until we reach the end which will be 'this'
+                nextParts = GetNextParts(parentResourceType, ref remaining);
 
-        private static ResourceType ChooseResourceType(string resourceTypeName, ResourceIdentifier parent)
-        {
-            if (resourceTypeName.Equals(ResourceGroupsLowerKey, StringComparison.OrdinalIgnoreCase))
-                return ResourceGroupResourceType;
+                //check the 4 resource types so we can extract the common identifier values
+                switch (nextParts.SpecialType)
+                {
+                    case SpecialType.ResourceGroup:
+                        ResourceGroupName = nextParts.ResourceName.ToString();
```

Comment:
Could the ToString() be called lazily to avoid an allocation until this property is needed by a caller?